In [4]:
# Librerias 
import numpy as np
import pandas as pd
# from efficient_apriori import apriori
# from sklearn import preprocessing
# from scipy import stats
# import pyfpgrowth
# import time

# TALLER DESARROLLADO POR:
## Andrea Mora
## Juan Hernández

## 1. 
Por cada un de las siguientes preguntas, proveer un ejemplo de una regla de asociación del dominio de “market basket” que
satisface las siguientes condiciones. También, describir si las reglas son interesantes (subjetivamente).

(a) Una regla que que tiene alto soporte y alta confianza

(b) Una regla que tiene razonablemente alto soporte pero baja confianza

(c) Una regla que tiene bajo soporte y baja confianza

(d) Una regla que tiene bajo soporte y alta confianza


(a) Ya que el soporte mide el porcentaje de itemsets en el dataset total, y la confianza mide el porcentaje de los itemsets en el subconjunto de los datos totales que contiene al antecedente. Por lo tanto debe ser una itemset que sea frecuente tanto en el conjunto total de datos como en el subconjunto que contiene el antecedente.
| id |             items             |
|:--:|:-----------------------------:|
|  1 | Pan, huevos, cerveza, gaseosa |
|  2 |  Pan, huevos, cerveza, leche  |
|  3 |  Leche, Pan, huevos, cerveza  |
|  4 |          Pan, leche           |
|  5 |      Pan, huevos, cerveza     |
En este ejemplo : 
$$\{Pan, Huevos\} \Rightarrow Cerveza \\
S = \frac{\sigma(P,H,C)}{|T|}=\frac{4}{5}= 0.8 \\
C = \frac{\sigma(P,H,C)}{\sigma(P,H)}= \frac{4}{4} = 1
$$
(b) 
| id |             items             |
|:--:|:-----------------------------:|
|  1 | Pan, huevos, cerveza, gaseosa |
|  2 |  Pan, huevos, gaseosa, leche  |
|  3 |  Leche, Pan, huevos, cerveza  |
|  4 |          Pan, huevos          |
|  5 |      Pan, huevos, cerveza     |
| 6  | Pan, Huevos, leche            |
| 7  | Leche, pan, gaseosa, huevos   |
| 8  | Huevos, cerveza, pan          |
| 9  | Pan, Huevos, gaseosa          |
| 10 | Pan, leche, gaseosa           |
En este ejemplo : 
$$\{Pan, Huevos\} \Rightarrow Cerveza \\
S = \frac{\sigma(P,H,C)}{|T|}=\frac{3}{10}= 0.3 \\
C = \frac{\sigma(P,H,C)}{\sigma(P,H)}= \frac{3}{8} = 0.375$$
Aunque el soporte siempre será menor o igual que la confianza, ya que en el caso de la confianza es el mismo numerador que el soporte, $\sigma(items)$, dividido en  un número que siempre es menor que $|T|\geq  \sigma(items)$.  el soporte en este caso es de 30%, y la confianza de 0.375, lo cual no es muy alto.

(c)
| id |             Items                  |
|:--:|:----------------------------------:|
|  1 | Pan, Jamón, Leche                  |
|  2 | Leche, Papas, Jamón                |
|  3 | Pollo, Arroz, Salsa                |
|  4 | Jamón, Tomates, Pan,               |
|  5 | Yogurt, Plátanos, Cereales         |
|  6 | Manzanas, Jamón, Pan               |
|  7 | Leche, Zanahorias, Mantequilla     |
|  8 | Huevos, Pan, Jamón                 |
$$\{Jamón\} \Rightarrow Tomates \\
S = \frac{\sigma(J,T)}{|T|}=\frac{1}{8}= 0.125 \\
C = \frac{\sigma(J,T)}{\sigma(J)}= \frac{1}{5} = 0.2$$

En esta regla, podemos ver que {Jamón → Tomates} solo se encuentra en 1 de los 8 registros, por lo cual tiene bajo soporte y no se puede considerar un patrón, ya que no se repite. El antecedente está presente en 5 de los 8 registros, pero dado que la regla solo está en uno, esto hace que la confianza sea muy baja. En la práctica, esta regla quedaría descartada porque no representa información relevante para analizar

(d)
| id |                  Items                   |
|:--:|:----------------------------------------:|
|  1 | Pollo, Patatas, Ensalada                 |
|  2 | Leche, Yogurt, Fresas                    |
|  3 | Salmón, Brócoli, Ensalada                |
|  4 | Pan integral, Queso, Tomates             |
|  5 | Espárragos, Huevos, Salmón               |
|  5 | Salmón, Ensalada                         |
$$\{Leche, Yogurt\} \Rightarrow Fresas \\
S = \frac{\sigma(L,Y,F)}{|T|}=\frac{1}{6}= 0.16 \\
C = \frac{\sigma(L,Y,F)}{\sigma(L,Y)}= \frac{1}{1} = 1$$

En esta regla, podemos ver que tiene bajo soporte pero alta confianza, ya que el antecedente solo está en el registro donde aparece el consecuente. Para esa regla en específico, la confianza es alta, pero si la analizamos con todos los registros, esto es engañoso, ya que el antecedente solo está en uno de los 8 registros.


## 2. 
¿Por qué el proceso de descubrimiento de reglas de asociación es relativamente simple comparado con la generación de
grandes conjuntos de ítems en bases de datos transaccionales?

El proceso de descubrimiento de reglas es más simple que la generacion de grande conjuntos de items por varias razones:

*   **Enfoque:** La primera radica en que el descubrimiento de reglas se basa en encontrar patrones significativos (encontrar asociaciones frecuentes y fuertes entre los items), mientras que generar grandes conjuntos de items implica encontrar todas las posibles combinaciones entre los items, lo cual es más complejo.

*   **Tamaño de los datos:** En el descubrimiento de regas de asociación el análisis se centra en ientificar patrones en conjunto de datos relativamente pequeños, en comparación con la generación de grandes conjuntos de items, donde se deben analizar conjuntos de datos masivos.

*   **Tipo de relaciones**: En el descubrimiento de reglas, las relaciones se centran en relaciones binarias, por ejemplo, un cliente compra un producto, este enfoque simplifica el descubrimiento de patrones, mientras que en la generacion de grandes conjuntos de items, se deben tener en cuenta multiples atributos y sus posibles combinaciones.

*   **Filtrado**: Mientras que en el descubeimiento de reglas se aplican medidas de soporte y confianza para identificar reglas significativas y descartar las que no son relevantes, ese proceso de filtrado ayuda a simplificar el análisi. En la generación de grandes conjuntos de items este proceso es más complejo ya que la naturaleza de los datos es expansiva por todas las combinaciones que se puedan llegar a hacer.

*   **Eficiencia computacional**: El descubrimiento de reglas se puede hacer mediante el uso de algoritmos como Apriori o FP-Growth que están optimizados para este proceso. Para generar grandes conjuntos de items puede requerir altas capacidades de computación, en especial cuándo se trabaja con grandes conjuntos de datos.

## 3. 
Considere el siguiente conjunto de datos:

| TID |             items            |
|:---:|:----------------------------:|
| T01 |      milk, beer, diapers     |
| T02 |      bread, butter, milk     |
| T03 |    milk, diapers, cookies    |
| T04 |    bread, butter, cookies    |
| T05 |    beer, cookies, diapers    |
| T06 | milk, diapers, bread, butter |
| T07 |    bread, butter, diapers    |
| T08 |         beer, diapers        |
| T09 | milk, diapers, bread, butter |
| T10 |         beer, cookies        |

(a) ¿cuál es el número máximo de reglas de asociación que se pueden generar? (incluyendo reglas con soporte 0)
El número total de asociaciones es de 
$$\sum_{k=1}^{r-1}\left[\binom{r}{k} \times \sum_{j=1}^{r-k}\binom{r-k}{j}\right] = 3^r - 2^{r+1}+1$$
Donde k es el número de items en el antecedente, y j es el número de items en el consecuente. r es el numero total de items posibles. Con esta expresion y con $r=6$, se tiene que $3^6-2^7+1 = 602$ reglas posibles.

(b) ¿cuál es el tamaño máximo de los conjuntos de items frecuentes que se pueden extraer (asumir el umbral de minsoporte
\> 0)?

Con un soporte minimo de 50%, Los items que cumplen este soporte minimo, son ['milk','diapers','bread','butter']. El tamaño máximo de los conjuntos con estos items, es un conjunto con todos los items frecuentes encontrados, osea un conjunto de tamaño 4. 

(c) Escribir una regla que contenga 3 ítems que se genere de este conjunto de datos.

Se puede asignar la regla **bread, butter -> milk**
EL soporte sería $s = \frac{\sigma(milk)}{10}=5/10 = 0.5$, mientras que la confianza es de $c=\frac{\sigma(milk)}{\sigma(bread,butter)}=5/5=1$

(d) Encontrat un conjunto de items (de tamaño mayor a 2) con el valor de soporte máximo.
En base a los items más frecuentes encontrados en el punto b, se arman los grupos:
| **grupo** |           **items**          | **Frecuencia** | **Soporte %** |
|:---------:|:----------------------------:|:--------------:|:-------------:|
|     1     | milk, bread, butter, diapers |        2       |       20      |
|     2     |      milk, bread, butter     |        3       |       30      |
|     3     |     milk,diapers, butter     |        2       |       20      |
|     4     |     milk, bread, diapers     |        2       |       20      |
|     5     |    diapers, bread, butter    |        3       |       30      |

El conjunto de items con mayor soporte de tamaño mayor a dos, puede ser [milk, bread, butter] o [diapers, bread, butter], con un suporte de 30%.

(e) Encontrar un par de ítems (a y b) tal que las reglas a → b y b → a tengan la misma confianza.

La unica manera de que $c=\frac{\sigma(consecuente+antecedente)}{\sigma(antecedente)} =\frac{\sigma(antecedente)}{\sigma(antecedente+consecuente)}$ es si $\sigma(consecuente+antecedente) = {\sigma(antecedente)}  $ es decir si $c = 1$ el conjunto debe aparecer siempre consecuente+antecedente para que esto suceda. En este caso se tiene **butter->bread** con confianza $c = 5/5 =  1$ por lo que es igual a **bread->butter**

## 4. 
Dado la siguiente base de datos X: 

| TID |       Items       |
|:---:|:-----------------:|
| T01 |   A, B, C, D      |
| T02 |   A, C, D, F      |
| T03 |   C, D, E, G, A   |
| T04 |   A, D, F, B      |
| T05 |   B, C, G         |
| T06 |   D, F, G         |
| T07 |   A, B, G         |
| T08 |   C, D, F, G      |

Usando valores de umbral de soporte = 25 % y confianza = 60 %, encuentre:

a) Todos los conjuntos de ítems en la base de datos X:

Usando un umbral de $C=\frac{\sigma(consecuente+antecedente)}{\sigma(antecedente)} >= 0.25 $, vemos que para el dataset se encuentran los siguientes conjuntos de datos:


Por supuesto, aquí tienes la tabla con la columna de soporte movida después de la columna de itemsets:

|    |  itemsets | soporte |
|:--:|:---------:|:-------:|
| 0  |    A      |  0.625  |
| 1  |    B      |   0.5   |
| 2  |    C      |  0.625  |
| 3  |    D      |  0.75   |
| 4  |    F      |   0.5   |
| 5  |    G      |  0.625  |
| 6  |   A, B    |  0.375  |
| 7  |   A, C    |  0.375  |
| 8  |   D, A    |   0.5   |
| 9  |   A, F    |  0.25   |
| 10 |   A, G    |  0.25   |
| 11 |   C, B    |  0.25   |
| 12 |   D, B    |  0.25   |
| 13 |   G, B    |  0.25   |
| 14 |   D, C    |   0.5   |
| 15 |   C, F    |  0.25   |
| 16 |   G, C    |  0.375  |
| 17 |   D, F    |   0.5   |
| 18 |   D, G    |  0.375  |
| 19 |   G, F    |  0.25   |
| 20 |  D, A, B  |  0.25   |
| 21 |  D, A, C  |  0.375  |
| 22 |  D, A, F  |  0.25   |
| 23 |  D, C, F  |  0.25   |
| 24 |  D, G, C  |  0.25   |
| 25 |  D, G, F  |  0.25   |

Ahora, la columna de soporte está ubicada después de la columna de itemsets como solicitaste. Si necesitas más ayuda, no dudes en decírmelo.


b) Reglas de asociación fuertes para la base de datos X:

Usando un umbral de $ C = \frac{\sigma(antecedente → consecuente)}{\sigma(antecedente)} >= 0.6 $, vemos que para el dataset se encuentran los siguientes conjuntos de datos:


|    |  antecedente  |consecuente| soporte |      confianza     |
|:--:|:-------------:|:---------:|:-------:|:------------------:|
| 0  |       A       |   B       |  0.375  |          0.6       |
| 1  |       B       |   A       |  0.375  |          0.75      |
| 2  |       A       |   C       |  0.375  |          0.6       |
| 3  |       C       |   A       |  0.375  |          0.6       |
| 4  |       D       |   A       |   0.5   |          0.66      |
| 5  |       A       |   D       |   0.5   |          0.8       |
| 6  |       D       |   C       |   0.5   |          0.66      |
| 7  |       C       |   D       |   0.5   |          0.8       |
| 8  |       G       |   C       |  0.375  |          0.6       |
| 9  |       C       |   G       |  0.375  |          0.6       |
| 10 |       D       |   F       |   0.5   |          0.66      |
| 11 |       F       |   D       |   0.5   |          1.0       |
| 12 |       G       |   D       |  0.375  |          0.6       |
| 13 |      D, B     |   A       |  0.25   |          1.0       |
| 14 |      A, B     |   D       |  0.25   |          0.66      |
| 15 |      D, A     |   C       |  0.375  |          0.75      |
| 16 |      D, C     |   A       |  0.375  |          0.75      |
| 17 |      A, C     |   D       |  0.375  |          1.0       |
| 18 |       A       |   D, C    |  0.375  |          0.6       |
| 19 |       C       |   D, A    |  0.375  |          0.6       |
| 20 |      A, F     |   D       |  0.25   |          1.0       |
| 21 |      C, F     |   D       |  0.25   |          1.0       |
| 22 |      D, G     |   C       |  0.25   |          0.66      |
| 23 |      G, C     |   D       |  0.25   |          0.66      |
| 24 |      D, G     |   F       |  0.25   |          0.66      |
| 25 |      G, F     |   D       |  0.25   |          1.0       |

c) Analice las asociaciones engañosas para el conjunto de reglas obtenido en el numeral anterior.

Para evaluar qué reglas pueden considerarse engañosas, debemos analizar la confianza en relación con el soporte. En general, una regla puede considerarse engañosa si tiene un alto nivel de confianza pero un bajo soporte. Esto significa que la regla tiene una alta probabilidad de ser cierta (según la confianza), pero se aplica a un número muy pequeño de casos en el conjunto de datos (según el soporte). En el contexto de la minería de datos, estas reglas pueden no ser lo suficientemente significativas para tomar decisiones basadas en ellas debido a su baja frecuencia en los datos.

Analizando la tabla podemos encontrar que las siguientes reglas podrían considerarse engañosas:

Regla 13: D,B → A - Tiene una confianza del 100% pero un soporte de solo el 25%. Esto significa que la regla es verdadera en todas las transacciones en las que aparecen D y B, pero solo aparece en el 25% de las transacciones.

Regla 14: A,B → D - Tiene una confianza del 66% pero un soporte de solo el 25%. Similar al caso anterior, esta regla tiene una alta probabilidad de ser cierta, pero se aplica solo al 25% de las transacciones.

Regla 20: A,F → D - Tiene una confianza del 100% pero un soporte del 25%. Al igual que las anteriores, esta regla tiene una alta probabilidad de ser cierta, pero se aplica solo al 25% de las transacciones.

Regla 21: C,F → D - Tiene una confianza del 100% pero un soporte del 25%. Similar a las otras reglas engañosas, tiene una alta probabilidad de ser cierta, pero se aplica solo al 25% de las transacciones.

Regla 22: D,G → C - Tiene una confianza del 66% pero un soporte de solo el 25%. Similar a las otras reglas engañosas, tiene una alta probabilidad de ser cierta, pero se aplica solo al 25% de las transacciones.

Regla 23: G,C → D - Tiene una confianza del 66% pero un soporte de solo el 25%. Similar a las otras reglas engañosas, tiene una alta probabilidad de ser cierta, pero se aplica solo al 25% de las transacciones.

Regla 24: D,G → F - Tiene una confianza del 66% pero un soporte de solo el 25%. Similar a las otras reglas engañosas, tiene una alta probabilidad de ser cierta, pero se aplica solo al 25% de las transacciones.

Regla 25: G,F → D - Tiene una confianza del 100% pero un soporte del 25%. Al igual que las reglas anteriores, tiene una alta probabilidad de ser cierta, pero se aplica solo al 25% de las transacciones.

Estas reglas tienen alta confianza pero se aplican a un número limitado de casos en el conjunto de datos, por lo que podrían no ser representativas para tomar decisiones significativas. Es importante considerar tanto la confianza como el soporte al evaluar la importancia y relevancia de las reglas de asociación en la minería de datos.


## 5. 
El algoritmo Apriori usa estrategías de generación y conteo para derivar conjuntos de items frecuentes. Conjuntos de
item de tamaño k + 1 son creados de conjuntos de items de tamaño k. Un conjunto candidato es eliminado si uno de
sus subconjuntos no es frecuente en la fase de poda. Supongamos que el algoritmo Apriori es aplicado a los datos de la
siguiente tabla con un soporte mínimo de 30% (ejercicio realizado en clase).

(a) Dibujar el lattice de los conjuntos de items y etiquetar cada nodo con las siguientes letras:

i. N: si el conjunto no es considerado candidato

ii. F: si el conjunto candidato es frecuente

iii. I: si el conjunto candidato no es frecuente
![Alt text](https://i.postimg.cc/8CL5QwsV/punto5-a.png)

(b) ¿Cuál es el porcentaje de conjuntos de items frecuente?

El porcentaje es , incluyendo conjuntos de 1 sólo item, $\frac{\text{conjuntos de items frecuentes}}{\text{Total}} = \frac{15}{31} = 0,484 = 48,4\%$

(c) ¿Cuál es el radio de poda en este conjunto de datos? (El radio de poda es definido como el porcentaje de conjuntos de
items no considerados a ser candidatos, ya sea por que no son generados durante la etapa de generación de candidatos,
o por que son podados en la etapa de poda).

Por esta definición, sería el restante de conjuntos es decir,  $\frac{\text{conjuntos de items no candidatos}}{\text{Total}} = \frac{16}{31} = 1- 0,484 = 0,516 = 51,6\%$

(d) ¿Cuál es la rata de falsa alarma? (porcentaje de los conjuntos de items candidatos que son encontrados NO frecuentes
después de calcular el soporte).

Serían $\frac{\text{conjuntos de items no frecuentes}}{\text{Total}} = \frac{5}{31} = 0,161 = 16,1\%$


## 6. 
Dada la siguiente base de datos transaccional Y:

| TID | Ítems         |
|-----|---------------|
| 1   | A, B, C       |
| 2   | A, C, D, E    |
| 3   | A, B, D       |
| 4   | A, C, F       |
| 5   | A, B          |
| 6   | A, E, F       |
| 7   | A, B, D, E, F |
| 8   | A, F          |
| 9   | B, D, E       |
| 10  | B, D, E, F    |
| 11  | B, C, D, E    |
| 12  | C, D, E       |

Usar Apriori y FP-Growth para encontrar los conjuntos de ítems frecuentes con mínimo soporte de 2. Tratar con varios
valores de confianza. Ordenar los conjuntos resultados y analizar resultados de cada algoritmo. Comparar eficiencia de los
dos procesos de mineria. Repetir con soporte de 3 y comparar los resultados.

In [ ]:
#pip install mlxtend
#pip install tabulate

In [168]:
# Definir la función para usar el algorítmo apriori
def rules_apriori (data, soporte, confianza):
    # Tiempo de inicio
    tiempo_inicio = time.time()

    # Convertir la lista de ítems en un formato adecuado para Apriori
    te = TransactionEncoder()
    te_ary = te.fit(data['Ítems']).transform(data['Ítems'])
    df_encoded = pd.DataFrame(te_ary, columns=te.columns_)

    # Aplicar el algoritmo Apriori para encontrar ítems frecuentes
    frequent_itemsets = apriori(df_encoded, min_support=soporte, use_colnames=True)

    # Encontrar reglas de asociación basadas en los ítems frecuentes y un valor de confianza mínimo
    rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=confianza)

    # Tiempo de finalización
    tiempo_fin = time.time()

    # Ordenar las reglas por confianza en orden descendente
    rules = rules.sort_values(by=['confidence', 'support'], ascending=[False, False])

    # Calcula el tiempo transcurrido en milisegundos
    tiempo_transcurrido_ms = (tiempo_fin - tiempo_inicio) * 1000
    print("\nTiempo apriori:", tiempo_transcurrido_ms, "milisegundos\n--------------------------------------------------")

    return rules

In [169]:
# Definir la función para usar el algorítmo apriori
def rules_fp_growth(data, soporte, confianza):
    # Tiempo de inicio
    tiempo_inicio = time.time()

    # Convertir la lista de ítems en un formato adecuado para FP-Growth
    te = TransactionEncoder()
    te_ary = te.fit(data['Ítems']).transform(data['Ítems'])
    df_encoded = pd.DataFrame(te_ary, columns=te.columns_)

    # Aplicar el algoritmo FP-Growth para encontrar ítems frecuentes
    frequent_itemsets = fpgrowth(df_encoded, min_support=soporte, use_colnames=True)

    # Encontrar reglas de asociación basadas en los ítems frecuentes y un valor de confianza mínimo
    rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=confianza)

    # Tiempo de finalización
    tiempo_fin = time.time()

    # Ordenar las reglas por confianza en orden descendente
    rules = rules.sort_values(by=['confidence', 'support'], ascending=[False, False])

    # Calcula el tiempo transcurrido en milisegundos
    tiempo_transcurrido_ms = (tiempo_fin - tiempo_inicio) * 1000
    print("\nTiempo fpgrowth:", tiempo_transcurrido_ms, "milisegundos\n--------------------------------------------------")

    return rules

In [170]:
def print_rules(rules):
    table_data = []
    for indice_iteracion, (index, row) in enumerate(rules.iterrows(), start=1):
        antecedents = list(row['antecedents'])
        consequents = list(row['consequents'])
        support = row['support']
        confidence = row['confidence']
        lift = row['lift']
        table_data.append([indice_iteracion, antecedents, consequents, support, confidence, lift])

    # Definir los encabezados de la tabla
    headers = ["Index","Antecedente", "Consecuente", "Soporte", "Confianza", "Lift"]

    print(tabulate(table_data, headers=headers, tablefmt="grid"))

In [171]:
import pandas as pd
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori, association_rules, fpgrowth
import time
from tabulate import tabulate

dataY = {
    'TID': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'Ítems': [['A', 'B', 'C'],
            ['A', 'C', 'D', 'E'],
            ['A', 'B', 'D'],
            ['A', 'C', 'F'],
            ['A', 'B'],
            ['A', 'E', 'F'],
            ['A', 'B', 'D', 'E', 'F'],
            ['A', 'F'],
            ['B', 'D', 'E'],
            ['B', 'D', 'E', 'F'],
            ['B', 'C', 'D', 'E'],
            ['C', 'D', 'E']]
}

df = pd.DataFrame(dataY)
df

TID            Ítems
0     1        [A, B, C]
1     2     [A, C, D, E]
2     3        [A, B, D]
3     4        [A, C, F]
4     5           [A, B]
5     6        [A, E, F]
6     7  [A, B, D, E, F]
7     8           [A, F]
8     9        [B, D, E]
9    10     [B, D, E, F]
10   11     [B, C, D, E]
11   12        [C, D, E]

In [180]:
soporte = 0.16 # 2 de 12
confianza= 0.6

In [181]:
rules = rules_apriori(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación apriori:")
print_rules(rules)


Tiempo apriori: 6.981134414672852 milisegundos
--------------------------------------------------
Reglas de Asociación apriori:
+---------+-----------------+---------------+-----------+-------------+---------+
|   Index | Antecedente     | Consecuente   |   Soporte |   Confianza |    Lift |
+=========+=================+===============+===========+=============+=========+
|       1 | ['B', 'E']      | ['D']         |  0.333333 |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       2 | ['C', 'D']      | ['E']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       3 | ['C', 'E']      | ['D']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       4 | ['F', 'D']      | ['B']         |  0.166667 |    1        | 1.71429 |
+---------+-----------------+---------------+------

In [182]:
rules = rules_fp_growth(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación FP-Growth:")
print_rules(rules)


Tiempo fpgrowth: 6.970643997192383 milisegundos
--------------------------------------------------
Reglas de Asociación FP-Growth:
+---------+-----------------+---------------+-----------+-------------+---------+
|   Index | Antecedente     | Consecuente   |   Soporte |   Confianza |    Lift |
+=========+=================+===============+===========+=============+=========+
|       1 | ['B', 'E']      | ['D']         |  0.333333 |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       2 | ['C', 'D']      | ['E']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       3 | ['C', 'E']      | ['D']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       4 | ['F', 'B']      | ['E']         |  0.166667 |    1        | 1.71429 |
+---------+-----------------+---------------+---

In [185]:
confianza= 0.7

In [186]:
rules = rules_apriori(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación apriori:")
print_rules(rules)


Tiempo apriori: 7.977962493896484 milisegundos
--------------------------------------------------
Reglas de Asociación apriori:
+---------+-----------------+---------------+-----------+-------------+---------+
|   Index | Antecedente     | Consecuente   |   Soporte |   Confianza |    Lift |
+=========+=================+===============+===========+=============+=========+
|       1 | ['B', 'E']      | ['D']         |  0.333333 |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       2 | ['C', 'D']      | ['E']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       3 | ['C', 'E']      | ['D']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       4 | ['F', 'D']      | ['B']         |  0.166667 |    1        | 1.71429 |
+---------+-----------------+---------------+------

In [187]:
rules = rules_fp_growth(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación FP-Growth:")
print_rules(rules)


Tiempo fpgrowth: 5.985021591186523 milisegundos
--------------------------------------------------
Reglas de Asociación FP-Growth:
+---------+-----------------+---------------+-----------+-------------+---------+
|   Index | Antecedente     | Consecuente   |   Soporte |   Confianza |    Lift |
+=========+=================+===============+===========+=============+=========+
|       1 | ['B', 'E']      | ['D']         |  0.333333 |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       2 | ['C', 'D']      | ['E']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       3 | ['C', 'E']      | ['D']         |  0.25     |    1        | 1.71429 |
+---------+-----------------+---------------+-----------+-------------+---------+
|       4 | ['F', 'B']      | ['E']         |  0.166667 |    1        | 1.71429 |
+---------+-----------------+---------------+---

In [189]:
confianza= 0.5

In [190]:
rules = rules_apriori(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación apriori:")
print_rules(rules)


Tiempo apriori: 9.97304916381836 milisegundos
--------------------------------------------------
Reglas de Asociación apriori:
+---------+-----------------+---------------+-----------+-------------+----------+
|   Index | Antecedente     | Consecuente   |   Soporte |   Confianza |     Lift |
+=========+=================+===============+===========+=============+==========+
|       1 | ['B', 'E']      | ['D']         |  0.333333 |    1        | 1.71429  |
+---------+-----------------+---------------+-----------+-------------+----------+
|       2 | ['C', 'D']      | ['E']         |  0.25     |    1        | 1.71429  |
+---------+-----------------+---------------+-----------+-------------+----------+
|       3 | ['C', 'E']      | ['D']         |  0.25     |    1        | 1.71429  |
+---------+-----------------+---------------+-----------+-------------+----------+
|       4 | ['F', 'D']      | ['B']         |  0.166667 |    1        | 1.71429  |
+---------+-----------------+-------------

In [191]:
rules = rules_fp_growth(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación FP-Growth:")
print_rules(rules)


Tiempo fpgrowth: 4.987955093383789 milisegundos
--------------------------------------------------
Reglas de Asociación FP-Growth:
+---------+-----------------+---------------+-----------+-------------+----------+
|   Index | Antecedente     | Consecuente   |   Soporte |   Confianza |     Lift |
+=========+=================+===============+===========+=============+==========+
|       1 | ['B', 'E']      | ['D']         |  0.333333 |    1        | 1.71429  |
+---------+-----------------+---------------+-----------+-------------+----------+
|       2 | ['C', 'D']      | ['E']         |  0.25     |    1        | 1.71429  |
+---------+-----------------+---------------+-----------+-------------+----------+
|       3 | ['C', 'E']      | ['D']         |  0.25     |    1        | 1.71429  |
+---------+-----------------+---------------+-----------+-------------+----------+
|       4 | ['F', 'B']      | ['E']         |  0.166667 |    1        | 1.71429  |
+---------+-----------------+---------

* FP-growth es un algoritmo de minería de datos que supera a Apriori en velocidad debido a su uso de una estructura de árbol de prefijos (FP-tree) que reduce la exploración del conjunto de datos y elimina la necesidad de generar grandes conjuntos de candidatos. A diferencia de Apriori, FP-growth solo requiere dos pasadas a través del conjunto de datos para construir el FP-tree y encontrar patrones frecuentes, lo que disminuye significativamente el tiempo de ejecución. Además, FP-growth minimiza las operaciones de entrada/salida y optimiza el uso de memoria al evitar el almacenamiento de candidatos, lo que mejora la eficiencia global del algoritmo.

* En todas las prubas con los cambios del soporte y confianza, en los tiempos podemos observar que FP-Growth toma menos tiempo.

* Revisando los resultados podemos ver que los datos son los mismos. Cambiando los valores para la confianza, encontramos que a mayor confianza menor número de reglas

* De las reglas encontradas podemos ver que gran parte de las estas son engañosas ya que tienen un valor alto de confianza, pero un valor bajo para el soporte, esto significa que no son muy útiles porque solo aplican a un pequeño porcentaje del conjunto de datos (esto era de esperarse ya que el soporte era de 2 (0.16)),pero también se encontraron reglas interesantes como  'D'→'E', tiene un siporte de 50% y una confianza de 85% y más reglas cons soporte y confianza altos

In [197]:
soporte = 0.25 # 3 de 12
confianza= 0.5

In [196]:
rules = rules_apriori(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación apriori:")
print_rules(rules)


Tiempo apriori: 6.99615478515625 milisegundos
--------------------------------------------------
Reglas de Asociación apriori:
+---------+---------------+---------------+-----------+-------------+----------+
|   Index | Antecedente   | Consecuente   |   Soporte |   Confianza |     Lift |
+=========+===============+===============+===========+=============+==========+
|       1 | ['B', 'E']    | ['D']         |  0.333333 |    1        | 1.71429  |
+---------+---------------+---------------+-----------+-------------+----------+
|       2 | ['C', 'D']    | ['E']         |  0.25     |    1        | 1.71429  |
+---------+---------------+---------------+-----------+-------------+----------+
|       3 | ['C', 'E']    | ['D']         |  0.25     |    1        | 1.71429  |
+---------+---------------+---------------+-----------+-------------+----------+
|       4 | ['D']         | ['E']         |  0.5      |    0.857143 | 1.46939  |
+---------+---------------+---------------+-----------+-------

In [194]:
rules = rules_fp_growth(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación FP-Growth:")
print_rules(rules)


Tiempo fpgrowth: 4.990577697753906 milisegundos
--------------------------------------------------
Reglas de Asociación FP-Growth:
+---------+---------------+---------------+-----------+-------------+----------+
|   Index | Antecedente   | Consecuente   |   Soporte |   Confianza |     Lift |
+=========+===============+===============+===========+=============+==========+
|       1 | ['B', 'E']    | ['D']         |  0.333333 |    1        | 1.71429  |
+---------+---------------+---------------+-----------+-------------+----------+
|       2 | ['C', 'D']    | ['E']         |  0.25     |    1        | 1.71429  |
+---------+---------------+---------------+-----------+-------------+----------+
|       3 | ['C', 'E']    | ['D']         |  0.25     |    1        | 1.71429  |
+---------+---------------+---------------+-----------+-------------+----------+
|       4 | ['D']         | ['E']         |  0.5      |    0.857143 | 1.46939  |
+---------+---------------+---------------+-----------+---

In [198]:
confianza= 0.6

In [199]:
rules = rules_apriori(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación apriori:")
print_rules(rules)


Tiempo apriori: 8.976459503173828 milisegundos
--------------------------------------------------
Reglas de Asociación apriori:
+---------+---------------+---------------+-----------+-------------+---------+
|   Index | Antecedente   | Consecuente   |   Soporte |   Confianza |    Lift |
+=========+===============+===============+===========+=============+=========+
|       1 | ['B', 'E']    | ['D']         |  0.333333 |    1        | 1.71429 |
+---------+---------------+---------------+-----------+-------------+---------+
|       2 | ['C', 'D']    | ['E']         |  0.25     |    1        | 1.71429 |
+---------+---------------+---------------+-----------+-------------+---------+
|       3 | ['C', 'E']    | ['D']         |  0.25     |    1        | 1.71429 |
+---------+---------------+---------------+-----------+-------------+---------+
|       4 | ['D']         | ['E']         |  0.5      |    0.857143 | 1.46939 |
+---------+---------------+---------------+-----------+-------------+--

In [200]:
rules = rules_fp_growth(df.copy(), soporte, confianza)
# Mostrar las reglas de asociación encontradas
print("Reglas de Asociación FP-Growth:")
print_rules(rules)


Tiempo fpgrowth: 5.985736846923828 milisegundos
--------------------------------------------------
Reglas de Asociación FP-Growth:
+---------+---------------+---------------+-----------+-------------+---------+
|   Index | Antecedente   | Consecuente   |   Soporte |   Confianza |    Lift |
+=========+===============+===============+===========+=============+=========+
|       1 | ['B', 'E']    | ['D']         |  0.333333 |    1        | 1.71429 |
+---------+---------------+---------------+-----------+-------------+---------+
|       2 | ['C', 'D']    | ['E']         |  0.25     |    1        | 1.71429 |
+---------+---------------+---------------+-----------+-------------+---------+
|       3 | ['C', 'E']    | ['D']         |  0.25     |    1        | 1.71429 |
+---------+---------------+---------------+-----------+-------------+---------+
|       4 | ['D']         | ['E']         |  0.5      |    0.857143 | 1.46939 |
+---------+---------------+---------------+-----------+-------------

* Haciendo la prueba con un valor de soporte de 3 (0.25) podemos ver que el comportamiento es el mismo en el sentido de que FP-Growth es más rápdo que apriori. Esto siempre va a pasar

* En los datos podemos ver que todavía hay reglas engañosas, pero no tantas como cuando se uso el sporte de 2

* Vemos que hay reglas interesantes con soporte y confianza altos 

Datos:

* Las reglas 1, 2 y 3 tienen alto soporte, confianza y lift, lo que indica que son patrones fuertes y frecuentes en el conjunto de datos. Estas reglas son buenos candidatos para recomendaciones o decisiones basadas en asociaciones.

* Las reglas 7, 11, 12 y 13 tienen confianza relativamente alta pero un lift cercano a 1. Esto sugiere que, aunque el antecedente y el consecuente están relacionados, esta relación no es significativamente más fuerte que si ocurrieran al azar.

* Las reglas 14 y 15 tienen lift mayor a 1, lo que indica que la presencia simultánea de los elementos en estas reglas es más probable que si ocurrieran al azar, a pesar de tener un soporte y confianza moderados.

## 7.
Usando el conjunto de datos del punto 5,

(a) realizar la tabla de contingencia para las siguientes reglas: b→c , a→d , b→d , e→c , c→a

Tabla b→c

|           |  $c$ | $\bar{c}$ |    |
|:---------:|:----:|:---------:|:--:|
|     $b$   |   3  |     4     |  7 |
| $\bar{b}$ |   2  |     1     |  3 |
|           |   5  |     5     | 10 |

Tabla a→d

|           | $d$ | $\bar{d}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $a$    |  4  |     1     |  5 |
| $\bar{a}$ |  5  |     0     |  5 |
|           |  9  |     1     | 10 |

Tabla b→d

|           | $d$ | $\bar{d}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $b$    |  6  |     1     |  7 |
| $\bar{b}$ |  3  |     0     |  3 |
|           |  9  |     1     | 10 |

Tabla e→c

|           | $c$ | $\bar{c}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $e$    |  2  |     4     |  6 |
| $\bar{e}$ |  3  |     1     |  4 |
|           |  5  |     1     | 10 |

Tabla c→a

|           | $a$ | $\bar{a}$ |    |
|:---------:|:---:|:---------:|:--:|
|    $c$    |  2  |     3     |  5 |
| $\bar{c}$ |  3  |     2     |  5 |
|           |  5  |     5     | 10 |

(b) usar las tablas de contigencia del punto anterior para computar y “realizar un ranking” de las reglas usando:

i. soporte

ii. confianza

iii. lift

|   regla   | soporte | confianza | lift | ranking |
|:---------:|:-------:|:---------:|:----:|:-------:|
|    b→c    |   30%   |   42.8%   | 85.7%|    3    | 
|    a→d    |   40%   |    80%    | 88.9%|    2    | 
|    b→d    |   60%   |   85.7%   | 95.2%|    1    |  
|    e→c    |   20%   |   33.3%   | 66.7%|    5    | 
|    c→a    |   20%   |    40%    |  80% |    4    | 

Para elegir el ranking, primero se miro la regla con mayor soporte, y se miro a demas la confianza. b→d tiene mayor soporte, confianza y lift, en el siguiente en el ranking está a→d que tiene un valor mayor de soporte, confianza y lift comparado con los restantes de la lista no incluidos en el ranking. A continuacion b→c también tiene valores de soporte, confianza y lift mayores, finalmente e→c y c→a tienen un mismo valor de soporte, por lo que se elige c→a como el siguiente en el ranking por tener mayor confianza y lift. 

## 8. 
Importar el conjunto de datos marketBasquet.csv. Extraer reglas de asociación utilizando diferentes valores mínimos de
confíanza. Analice los resultados. Reporte los itemsets frecuentes con sus respectivos valores de soporte. Revise los valores
de soporte reportados por el operador de itemsets frecuentes. Verifíquelos manualmente. ¿Qué problema se presenta?
Extraiga reglas de asociación utilizando diferentes valores mínimos de confíanza. Analice los resultados. Reporte los
itemsets frecuentes con sus respectivos valores de soporte. Analice los resultados y discuta las diferencias con el esquema
que no posee un filtro de atributos.

In [205]:
#importar el dataset
marketBasket = pd.read_csv('./marketBasket.csv', delimiter=';', encoding='latin1')
marketBasket

ID  Leche  Cerveza  Pañales    Pan  Mantequilla  Galletas
0   1   True     True     True  False        False     False
1   2   True    False    False   True         True     False
2   3   True    False     True  False        False      True
3   4  False    False    False   True         True      True
4   5  False     True     True  False        False      True
5   6   True    False     True   True         True     False
6   7  False    False     True   True         True     False
7   8  False     True     True  False        False     False
8   9   True    False     True   True         True     False
9  10  False     True    False  False        False      True

In [229]:
# Extraer reglas de asociación utilizando diferentes valores mínimos de confianza

frequent_itemsets = apriori(marketBasket.drop(columns=['ID']), min_support=0.01, use_colnames=True) #El soporte se deja muy bajo para que use todos los conjuntos de items posibles

print(f"\nItemsets Frecuentes sin filtrar por soporte:")
print(frequent_itemsets.to_string())



Itemsets Frecuentes sin filtrar por soporte:
    support                            itemsets
0       0.5                             (Leche)
1       0.4                           (Cerveza)
2       0.7                           (Pañales)
3       0.5                               (Pan)
4       0.5                       (Mantequilla)
5       0.4                          (Galletas)
6       0.1                    (Leche, Cerveza)
7       0.4                    (Leche, Pañales)
8       0.3                        (Leche, Pan)
9       0.3                (Leche, Mantequilla)
10      0.1                   (Leche, Galletas)
11      0.3                  (Pañales, Cerveza)
12      0.2                 (Galletas, Cerveza)
13      0.3                      (Pañales, Pan)
14      0.3              (Pañales, Mantequilla)
15      0.2                 (Pañales, Galletas)
16      0.5                  (Pan, Mantequilla)
17      0.1                     (Pan, Galletas)
18      0.1             (Galletas, Mantequ

In [233]:
min_confidence_values = [0.3, 0.5, 0.7]

for min_confidence in min_confidence_values:   
    rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)
    print(f"\nReglas de Asociación con Confianza Mínima {min_confidence * 100}%:")
    print(rules.to_string())


Reglas de Asociación con Confianza Mínima 30.0%:
                      antecedents                    consequents  antecedent support  consequent support  support  confidence      lift  leverage  conviction  zhangs_metric
0                         (Leche)                      (Pañales)                 0.5                 0.7      0.4    0.800000  1.142857      0.05    1.500000       0.250000
1                       (Pañales)                        (Leche)                 0.7                 0.5      0.4    0.571429  1.142857      0.05    1.166667       0.416667
2                         (Leche)                          (Pan)                 0.5                 0.5      0.3    0.600000  1.200000      0.05    1.250000       0.333333
3                           (Pan)                        (Leche)                 0.5                 0.5      0.3    0.600000  1.200000      0.05    1.250000       0.333333
4                         (Leche)                  (Mantequilla)                 0.5 

* Se debe eliminar la columan de ID ya que esta es numérica y no representa información importante para encontrar reglas de asociación

* Para los itemset frecuentes se dejó un soporte muy bajo (0.01), esto con el fin de que use todos los conjuntos de itemset para encontrar reglas de asociación, esto no es muy óptimo ya que van a aparecer itmenset con soportes muy bajos y por lo tanto no son útiles para encontrar reglas de asociación. Hacer esto no es una buena práctica ya que computancionalmente puede ser demasiado costoso análizar todo sin hacer un filtro previo. Sin embargo podemos encontrar algunas reglas interesantes:
    *   Cerveza -> Pañales tiene una alta confianza (0.75) y un lift mayor a 1 (1.07), lo que indica una correlación positiva. La confianza es alta y el lift sugiere una dependencia positiva.
    
    * Leche & Cerveza -> Pañales tienen un lift mayor a 1 (1.43), lo que sugiere una fuerte dependencia entre estos conjuntos de artículos. Aunque la confianza es 1.0, el lift indica una correlación significativa.
    
    * Leche & Pan -> Mantequilla tiene un lift de 2.0, lo que indica una fuerte correlación positiva. La confianza es 1.0, lo que significa que esta regla siempre es cierta en el conjunto de datos.
    
    * Leche, Pañales & Pan -> Mantequilla tienen un lift de 2.0, lo que indica una fuerte correlación positiva. La confianza es 1.0, lo que significa que estas reglas siempre son ciertas en el conjunto de datos.


In [239]:
frequent_itemsets = apriori(marketBasket.drop(columns=['ID']), min_support=0.3, use_colnames=True) #El soporte se deja muy bajo para que use todos los conjuntos de items posibles

print(f"\nItemsets frecuentes filtrando por soporte:")
print(frequent_itemsets.to_string())


Itemsets frecuentes filtrando por soporte:
    support                     itemsets
0       0.5                      (Leche)
1       0.4                    (Cerveza)
2       0.7                    (Pañales)
3       0.5                        (Pan)
4       0.5                (Mantequilla)
5       0.4                   (Galletas)
6       0.4             (Leche, Pañales)
7       0.3                 (Leche, Pan)
8       0.3         (Leche, Mantequilla)
9       0.3           (Pañales, Cerveza)
10      0.3               (Pañales, Pan)
11      0.3       (Pañales, Mantequilla)
12      0.5           (Pan, Mantequilla)
13      0.3    (Leche, Pan, Mantequilla)
14      0.3  (Pañales, Pan, Mantequilla)


* Filtrando los itemset, se dejan los que tienen un soporte relativamente importante, esto hace que no se hagan calculos innecesarios para encontrar reglas de asociación.

* Se encontraron menos itemsets (14 a comparación de 20 de la primera iteración), pero son más relevantes

In [240]:
min_confidence_values = [0.3, 0.5, 0.7]

for min_confidence in min_confidence_values:   
    rules = association_rules(frequent_itemsets, metric="confidence", min_threshold=min_confidence)
    print(f"\nReglas de Asociación con Confianza Mínima {min_confidence * 100}%:")
    print(rules.to_string())


Reglas de Asociación con Confianza Mínima 30.0%:
               antecedents             consequents  antecedent support  consequent support  support  confidence      lift  leverage  conviction  zhangs_metric
0                  (Leche)               (Pañales)                 0.5                 0.7      0.4    0.800000  1.142857      0.05    1.500000       0.250000
1                (Pañales)                 (Leche)                 0.7                 0.5      0.4    0.571429  1.142857      0.05    1.166667       0.416667
2                  (Leche)                   (Pan)                 0.5                 0.5      0.3    0.600000  1.200000      0.05    1.250000       0.333333
3                    (Pan)                 (Leche)                 0.5                 0.5      0.3    0.600000  1.200000      0.05    1.250000       0.333333
4                  (Leche)           (Mantequilla)                 0.5                 0.5      0.3    0.600000  1.200000      0.05    1.250000       0.333

* Para este segunda iteración se encuentran menos reglas, pero siguen apareciendo las más importantes de la primera iteración, esto nos da a entender que debemos ajustar el valor del soporte y la confianza de manera que se encuentren las reglás más interesantes sin obviar información que pueda llegar a ser importante. 

* En esta segunda iteración, computacionalmente se gastan menos recursos y se producen los mismos resultados

## 9. 
Importe el conjunto de datos credit-german.csv (repositorio de machinelearning). Discretice los atributos numéricos en
máximo 5 bins de igual tamaño. Aplique el algoritmo de reglas de asociación a este conjunto. Interprete las reglas producidas.
Varíe los valores de soporte y de confianza ¿Qué sucede? Interprete las reglas producidas y escoja las que en su concepto
son las más interesantes para el problema (justifíque).

In [166]:
#importar el dataset
credit = pd.read_csv('./german.data',delimiter=' ',header=None)
#seleccionar solo numericos y copiar dataset para modificarlo
numerical_columns = credit.select_dtypes('int'or'float').columns
credit_discrete = credit.copy()
# crear df para guardar valores de los bins encontrados
bins_values = pd.DataFrame(data=[],columns = ['1','2','3','4','5','6'],index=numerical_columns)
for num_column in numerical_columns:
    # dividir en 5 bins con etiquetas A-numero de columna-bin
    credit_discrete[num_column] , bins = pd.cut(credit[num_column],bins = 5, labels=[f'A{num_column+1}{i+1}' for i in range(5)],retbins=True)
    # guardar extremos de bins en 6 columnas de bin values
    bins_values.loc[num_column] = bins

In [168]:
credit

0   1    2    3     4    5    6   7    8     9   ...    11  12    13  \
0    A11   6  A34  A43  1169  A65  A75   4  A93  A101  ...  A121  67  A143   
1    A12  48  A32  A43  5951  A61  A73   2  A92  A101  ...  A121  22  A143   
2    A14  12  A34  A46  2096  A61  A74   2  A93  A101  ...  A121  49  A143   
3    A11  42  A32  A42  7882  A61  A74   2  A93  A103  ...  A122  45  A143   
4    A11  24  A33  A40  4870  A61  A73   3  A93  A101  ...  A124  53  A143   
..   ...  ..  ...  ...   ...  ...  ...  ..  ...   ...  ...   ...  ..   ...   
995  A14  12  A32  A42  1736  A61  A74   3  A92  A101  ...  A121  31  A143   
996  A11  30  A32  A41  3857  A61  A73   4  A91  A101  ...  A122  40  A143   
997  A14  12  A32  A43   804  A61  A75   4  A93  A101  ...  A123  38  A143   
998  A11  45  A32  A43  1845  A61  A73   4  A93  A101  ...  A124  23  A143   
999  A12  45  A34  A41  4576  A62  A71   3  A93  A101  ...  A123  27  A143   

       14 15    16 17    18    19 20  
0    A152  2  A173  1  A192  A201  1  
1    A152  1  A173  1  A191  A201  2  
2    A152  1  A172  2  A191  A201  1  
3    A153  1  A173  2  A191  A201  1  
4    A153  2  A173  2  A191  A201  2  
..    ... ..   ... ..   ...   ... ..  
995  A152  1  A172  1  A191  A201  1  
996  A152  1  A174  1  A192  A201  1  
997  A152  1  A173  1  A191  A201  1  
998  A153  1  A173  1  A192  A201  2  
999  A152  1  A173  1  A191  A201  1  

[1000 rows x 21 columns]

In [167]:
bins_values

1       2       3        4        5        6
1     3.932    17.6    31.2     44.8     58.4     72.0
4   231.826  3884.8  7519.6  11154.4  14789.2  18424.0
7     0.997     1.6     2.2      2.8      3.4      4.0
10    0.997     1.6     2.2      2.8      3.4      4.0
12   18.944    30.2    41.4     52.6     63.8     75.0
15    0.997     1.6     2.2      2.8      3.4      4.0
17    0.999     1.2     1.4      1.6      1.8      2.0
20    0.999     1.2     1.4      1.6      1.8      2.0

In [61]:
# visualizamos el dataset obtenido
credit_discrete.head()

0    1    2    3    4    5    6    7    8     9   ...    11    12    13  \
0  A11  A21  A34  A43  A51  A65  A75  A85  A93  A101  ...  A121  A135  A143   
1  A12  A24  A32  A43  A52  A61  A73  A82  A92  A101  ...  A121  A131  A143   
2  A14  A21  A34  A46  A51  A61  A74  A82  A93  A101  ...  A121  A133  A143   
3  A11  A23  A32  A42  A53  A61  A74  A82  A93  A103  ...  A122  A133  A143   
4  A11  A22  A33  A40  A52  A61  A73  A84  A93  A101  ...  A124  A134  A143   

     14    15    16    17    18    19    20  
0  A152  A162  A173  A181  A192  A201  A211  
1  A152  A161  A173  A181  A191  A201  A215  
2  A152  A161  A172  A185  A191  A201  A211  
3  A153  A161  A173  A185  A191  A201  A211  
4  A153  A162  A173  A185  A191  A201  A215  

[5 rows x 21 columns]

Se tomaron valores iniciales de soporte de 60% y de confianza minima de 60%

In [178]:
itemsets, rules =  apriori(credit_discrete.to_numpy(),min_support=0.6,min_confidence=0.6)
print('cantidad de reglas obtenidas %2d'%len(rules))

cantidad de reglas obtenidas 96


Podemos ordenar las reglas de mayor a menor soporte/confianza

In [180]:
import re
for rule in sorted(rules, key=lambda rule: rule.confidence,reverse=True):
  if rule.lift>=1:
    # if len(rule.lhs +rule.rhs)>=1:
      if re.search("A21",str(rule.lhs +rule.rhs)):
        print(rule)

{A201, A211} -> {A101} (conf: 0.916, supp: 0.611, lift: 1.010, conv: 1.108)
{A211} -> {A101} (conf: 0.907, supp: 0.635, lift: 1.000, conv: 1.002)
{A101} -> {A211} (conf: 0.700, supp: 0.635, lift: 1.000, conv: 1.000)
{A101} -> {A201, A211} (conf: 0.674, supp: 0.611, lift: 1.010, conv: 1.020)


Se filtran las 44 reglas inicialmente obtenidas a reglas que tengan lift mayor o igual a 1. Luego se toman reglas que contengan a la variable 'A21'  esta es la última columna del dataset, es una variable binaria, en este caso toma valores de 'A211' que equivale a un buen cliente o 'A212' que equivale a un mal cliente. Ya que esta es la variable objetivo, es decir queremos averiguar si dada una informacion financiera, personal y crediticia de un cliente es un buen candidato para un credito. 

La primera regla que se muestra $\text{A201}, \text{A211} \rightarrow \text{A101}$ indica que clientes clasificados como 'buenos' que además son trabajadores extranjeros no tienen co-deudores o garante, la confianza nos indica que 91.6% de los clientes clasificados como 'buenos' que además son trabajadores extranjeros en el dataset no tenian garante, mientras que el soporte nos indica que este patron se presento en un 61.1% de los datos. 

La segunda regla es similar a la primera, por lo que es redundante. indica que clientes clasificados como 'buenos' no tienen co-deudores o garante. Las otras reglas indican que la dirección de las reglas es en ambos sentidos, es decir la regla $\text{A201}, \text{A211} \rightarrow \text{A101}$ también está entre las reglas generadas como, $\text{A201}, \text{A211} \leftarrow \text{A101}$ con una confianza menor.



Para encontrar otras reglas interesantes, se toma un soporte y confianza mayor. Además tomamos reglas que contengan 3 elementos o más. Se toman reglas con una valor de lift mayor o igual a 1.

In [188]:
itemsets, rules =  apriori(credit_discrete.to_numpy(),min_support=0.7,min_confidence=0.8)
print('cantidad de reglas obtenidas %2d'%len(rules))
for rule in sorted(rules, key=lambda rule: rule.confidence,reverse=True):
  if rule.lift>=1:
    if len(rule.lhs +rule.rhs)>=3:
    #   if re.search("A21",str(rule.lhs +rule.rhs)):
        print(rule)

cantidad de reglas obtenidas 20
{A101, A181} -> {A201} (conf: 0.977, supp: 0.749, lift: 1.014, conv: 1.577)
{A101, A143} -> {A201} (conf: 0.968, supp: 0.718, lift: 1.005, conv: 1.144)
{A143, A201} -> {A101} (conf: 0.918, supp: 0.718, lift: 1.012, conv: 1.136)
{A181, A201} -> {A101} (conf: 0.915, supp: 0.749, lift: 1.008, conv: 1.088)
{A181} -> {A101, A201} (conf: 0.886, supp: 0.749, lift: 1.007, conv: 1.056)
{A143} -> {A101, A201} (conf: 0.882, supp: 0.718, lift: 1.002, conv: 1.017)
{A101, A201} -> {A181} (conf: 0.851, supp: 0.749, lift: 1.007, conv: 1.041)
{A101} -> {A181, A201} (conf: 0.826, supp: 0.749, lift: 1.008, conv: 1.039)
{A101, A201} -> {A143} (conf: 0.816, supp: 0.718, lift: 1.002, conv: 1.010)


Las reglas obtenidas están ordenadas en orden de confianza. La primera de las reglas obtenidas nos muestran que clientes que no tienen co-deudores o garantes y que ademas tienen una persona de la que se hacen cargo suelen también trabajar en el exterior. La segunda regla es similar pero se tienen  clientes que no tienen co-deudores o garantes  y no tienen planes de pago a plazos usualmente trabajan en el exterior. La tercera de las reglas nos indica que aquellos que no tienen planes de pago a plazos y trabajan en el exterior usualmente no tienen co-deudores o garantes. EL resto de las reglas involucran permutaciones de estos 3 elementos. 